#166272351 List Users functionality

GerrardE opened the pull request:
#### What does this PR do?
Implements the list users functionality for the application

#### Description of Task to be completed?
- Write tests
- Define api method in UserController
- Create endpoint `GET /users`
- Protect route

#### How should this be manually tested?
1. clone this repo
2. run "npm install"
3. run "npm test"

#### Any background context you want to provide?
`ALGORITHM`

Step 1: Start
Step 2: Take a new branch out of the develop
Step 3: Push to the remote repository
Step 4: Checkout of the local branch and pull to develop then checkout to the same branch
Step 5: Create database seeders for users and profiles,
Step 6: Write tests for the get route,
Step 7: Define the API method in existing UserController,
Step 8: Define the route in the routes folder
Step 9: Run tests and ensure they pass
Step 10: Push to the remote repository
Step 11: Stop

#### What are the relevant pivotal tracker stories?
[#166272351](https://www.pivotaltracker.com/story/show/166272351)

#### Screenshots (if appropriate)
<img width="1071" alt="Screenshot 2019-06-12 at 12 37 09 AM" src="https://user-images.githubusercontent.com/22410701/59313859-85e1af80-8caa-11e9-9ee4-430d9ccf9fb3.png">


#### Questions:
N/A

GerrardE committed:
ft-[166272351]: Implement user-profile listing

- Define api method in UserController
- Define route

[Delivers #166272351]

GerrardE committed:
ft-[166272351]: Implement user-profile listing

- Define api method in UserController
- Define route

[Delivers #166272351]

GerrardE committed:
ft-[166272351]: Implement user-profile listing

- Define api method in UserController
- Define route

[Delivers #166272351]

GerrardE committed:
ft-[166272351]: Implement user-profile listing

- Write tests and ensure they fail
- Define api method in UserController
- Define route

[Delivers #166272351]

GerrardE committed:
ft-[166272351]: Implement user-profile listing

- Write tests and ensure they fail
- Define api method in UserController
- Define route

[Delivers #166272351]

GerrardE committed:
ft-[166272351]: Implement user-profile listing

- Refactor test files

[Delivers #166272351]

fidelisojeah commented on server/tests/users.test.js:
```diff
@@ -1,14 +1,21 @@
 
 describe('TESTS TO SIGNUP A USER', () => {
+  before(async () => {
+    userToken = await generateToken({
```
This works

However, I don't want you to hardcode the user values.

in the `before` hook:

- You should generate a `new user` (Using `faker`)
- You should generate the `userToken` (as you have currently done but using the user details created from `faker`

GerrardE replied:
I studied `faker` docs and found how to implement the user generation just like you pointed out. the changes have been implemented in the new pull request and the test works just fine.

fidelisojeah replied:
it seems you only partially understand what I meant.

- Create a NEW USER in this `before` hook using `faker`
- generate token from that NEW USER

As it stands, when we merge #20 and this ticket, the system would run into problems

GerrardE replied:
Noted

fidelisojeah commented on server/controllers/users.js:
```diff
@@ -150,6 +151,41 @@ class UserController {
+        ]
+      });
+
+      if (isEmpty(users)) {
```
an empty list of users is still a valid request.

This should not be here

GerrardE replied:
I note this feedback

fidelisojeah commented on server/controllers/users.js:
```diff
@@ -150,6 +151,41 @@ class UserController {
+        .send({
+          status: 'success',
+          message: 'Users and corresponding profiles retrieved successfully',
+          users
```
So you would basically allow all users see the hashed passwords of all other users (and some other sensitive details) ?

GerrardE replied:
Before now I did not know how to perform exclude queries in sequelize. So I studied the querying section in their docs and found how to exclude columns in query attributes. I implemented it and was able to exclude sensitive information.

fidelisojeah replied:
also, return users as `payload`

```javascript
...
payload: users
});
```

GerrardE committed:
ft-[166272351]: Implement user-profile listing

- Refine query results for users and profiles
- Refactor test file

[Delivers #166272351]

GerrardE committed:
ft-[166272351]: Implement user-profile listing

- Refine query results for users and profiles

[Delivers #166272351]

fidelisojeah commented on server/controllers/users.js:
```diff
@@ -150,6 +150,41 @@ class UserController {
+  static async getUsers(req, res, next) {
+    try {
+      const users = await User.findAll({
+        attributes: { exclude: ['createdAt', 'updatedAt', 'password', 'email'] },
```
please change to include only attributes needed.

I see some unneeded attributes returned.

same thing for the profile model. (there's no need including `userId` in the profile aspect for instance

fidelisojeah commented on server/helpers/isEmpty.js:
```diff
@@ -0,0 +1,13 @@
+ * @param {string} value
+ * @return {(error|bool)} returns error or true
+ */
+const isEmpty = value => (
```
is this still used?
no? remove it

GerrardE replied:
Noted/Done

adex001 commented on package.json:
```diff
@@ -37,6 +37,7 @@
     "debug": "^4.1.1",
     "dotenv": "^8.0.0",
     "express": "^4.17.1",
+    "faker": "^4.1.0",
```
In my opinion, I think faker works to generate fake data and might not be really useful in our project. Why don't you add it under the devDependencies? I think it would be better there.

GerrardE replied:
Thank you

adex001 reviewed (changes requested):
Nice job. you can go ahead to implement the changes I suggested.

GerrardE committed:
ft-[166272351]: Implement user-profile listing

- Refine query results for users and profiles
- Refine test file

[Delivers #166272351]

adex001 commented:
Nice job so far!

fidelisojeah commented on server/controllers/users.js:
```diff
@@ -150,6 +150,41 @@ class UserController {
+  static async getUsers(req, res, next) {
+    try {
+      const users = await User.findAll({
+        attributes: { exclude: ['createdAt', 'updatedAt', 'password', 'email', 'middlename'] },
```
This would be the 3rd time I'm leaving this feedback

I don't want an exclude

use the attributes field as inclusive (this way, it excludes the fields you have not specified).


```javascript
...
attributes: [ 'field 1', 'field 2', 'field 3' ]
...
```

what you have, if the `User` model is modified and more sensitive field are added, you are exposing those fields to the world.

Please always consult documentation of libraries you are making use of.

GerrardE replied:
On it now

GerrardE replied:
I have reviewed the docs as you recommended, and I studied the use of `include` like you requested before implementing this new feedback.

GerrardE committed:
ft-[166272351]: Implement user-profile listing
- Refine query results for users and profiles
- Refactor test file

[Delivers #166272351]

GerrardE committed:
ft-[166272351]: Implement user-profile listing

- Update test files and resolve conflicts

[Delivers #166272351]

GerrardE committed:
ft-[166272351]: Implement user-profile listing

- Update test files and resolve conflicts

[Delivers #166272351]

GerrardE committed:
ft-[166272351]: Implement user-profile listing

- Update database querying method

[Delivers #166272351]

fidelisojeah commented on server/controllers/users.js:
```diff
@@ -151,13 +151,46 @@ class UserController {
+  */
+  static async getUsers(req, res, next) {
+    try {
+      const users = await User.findAll({
```
this should include

```suggestion
      const users = await User.findAll({
        ...
        where: {
          active: true
        }
      ...
```

we don't need to return the `active` label

GerrardE replied:
Resolved changes

fidelisojeah commented on server/app.js:
```diff
@@ -20,7 +20,7 @@ app.use(bodyParser.urlencoded({
 }));
 
-app.use('/api/v1/auth', authRoutes);
+app.use('/api', routes);
```
please modify this, pull from develop and do the right thing

GerrardE replied:
Changes updated and merge conflicts resolved

@fidelisojeah

fidelisojeah commented on server/routes/api/users.js:
```diff
@@ -10,5 +10,6 @@ userRoutes.put('/user', trim, UserController.updateUser);
 userRoutes.post('/login', trim, UserController.login);
```
move your `userRoutes.get('/users', Token.authorize, UserController.getUsers);` to the `userRoutes` file.

emmygozi commented:
Looks good to me. I can see the requested changes implemented.

GerrardE commented:
> Nice job so far!

@adex001 kindly approve

GerrardE committed:
ft-[166272351]: Implement user-profile listing

- Update database querying method

[Delivers #166272351]

GerrardE committed:
ft-[166272351]: Implement user-profile listing

- Update database querying method

[Delivers #166272351]

GerrardE committed:
ft-[166272351]: Implement user-profile listing

- Write more test to cover more edge cases

[Delivers #166272351]

GerrardE committed:
ft-[166272351]: Implement user-profile listing

- Write more test to cover more edge cases

[Delivers #166272351]

emmygozi reviewed (dismissed):
Timely implementing of changes. Good job.

fidelisojeah commented on server/tests/users.test.js:
```diff
@@ -1,13 +1,29 @@
+
+let authToken;
 
 describe('TESTS TO SIGNUP A USER', () => {
```
check name of test suite

fidelisojeah reviewed (commented):
Good job Implement minot changes

fidelisojeah commented on server/tests/users.test.js:
```diff
@@ -1,13 +1,29 @@
 
 describe('TESTS TO SIGNUP A USER', () => {
+  before(async () => {
+    const newUser = await User.create({
```
move user creation into the factory (and call the needed functions from here)

so
```javascript
User.create({
...
```
would be located in factory as a function called `createTestUser`

GerrardE committed:
ft-[166272351]: Implement user-profile listing

- Refactor user factory

[Delivers #166272351]

GerrardE committed:
ft-[166272351]: Implement user-profile listing

- Refactor user factory

[Delivers #166272351]

GerrardE committed:
ft-[166272351]: Implement user-profile listing

- Refactor test suite

[Delivers #166272351]

adex001 reviewed (approved):
Cool job
